### ISSUE TITLE
[BUG] UFW/Docker

### ISSUE THREAD

**Issue by mutech** [docker-mailserver/docker-mailserver#3151]
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

Documentation or setup.sh

### What happened and when does this occur?

Docker configures the host firewall to support exposed ports.

While this is nice, it conflicts with UFW, which does not show the firewall rules docker inserts. As a result, the output of `ufw status` looks like this:

```BASH
Status: active

To                         Action      From
--                         ------      ----
22/tcp on eth0             ALLOW       Anywhere   
```

However, all exposed ports of docker-mailserver are actually allowed.

This is of course not a bug in docker-mailserver but an inconsistency between ufw and docker. But since you mentioned in your documentation that the firewall has to be set up to support mailserver and actually use ufw as an example, you should instead mention this problem or solve it, for example using:
  * https://github.com/chaifeng/ufw-docker

### What did you expect to happen?

I would expect docker not to do it's own firewalling and if a project is based on docker and proposes to use ufw to open the firewall to support the container, then it should mention this issue.

My proposal is to prominently mention that docker does this kind of firewall setup on Linux and conflicts with tools like ufw.

As it stands now, users reading your instructions opening a subset of ports using ufw will almost certainly miss that all ports exposed through port mappings are automagically also accessible from the outside bypassing ufw.

### How do we replicate the issue?

1. Install a base ubuntu (or any distro supporting ufw)
2. Install docker-ce using the docker APT repository
3. Configure ufw to deny incoming traffic by default (opening whatever is otherwise needed)
4. Install docker-mailserver normally, but don't do any specific ufw setup for mail-related ports
5. Verify that all exposed ports are accessible despite no specific firewall setup 
6. Verify that ufw shows no rules allowing access to mail-related ports.

Looking at `iptables -t filter -L`  shows the rules docker inserts. 

### DMS version

v11.3.0

### What operating system is DMS running on?

Linux

### Which operating system version?

Ubuntu 22.04

### What instruction set architecture is DMS running on?

AMD64 / x86_64

### What container orchestration tool are you using?

Docker

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [X] I am rather experienced with docker
- [ ] I am inexperienced with mail servers
- [X] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [X] I am rather comfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by polarathene** [comment]
The only mention of UFW is in this [Examples - Basic Installation](https://docker-mailserver.github.io/docker-mailserver/edge/examples/tutorials/basic-installation/#a-detailed-example) docs page (_[earlier 11.3 version](https://docker-mailserver.github.io/docker-mailserver/v11.3/examples/tutorials/basic-installation/) for a static reference_)  with the "Detailed Example" section starting with clear communication of:

> This is a community contributed guide. Please let us know via a Github Issue if you're having any difficulty following the guide so that we can update it.

This was contributed by the community back when we had a Github Wiki with no review process. It's had some minor revisions since.

You are welcome to improve it. There is an edit button on the docs page at the start of the page content, it'll lead you [here](https://github.com/docker-mailserver/docker-mailserver/edit/master/docs/content/examples/tutorials/basic-installation.md). It's mostly just plain markdown text, but you can add a [`!!! caution "Some title"` admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#supported-types) to better inform users about this issue.

There's only a few maintainers for the project that volunteer their time, and unfortunately the docs don't get much love these days as we're rather stretched.

**Comment by mutech** [comment]
I hope I didn't come across as complaining about this, would have been unintentional. I was just shocked about what docker is doing and how and I thought that this might be relevant for people who operate a mailserver on a vps/ds on the internet. Considering that it takes not even a minute to fill the fail2ban jail with 20+ ips after installing a fresh instance, I got rather paranoid regarding firewall settings. But that was not meant as critique or pushing in your direction. You're doing a great job.

I actually don't know where I read about ufw, I just remembered that while I was looking for whatever info I needed I also read about firewall setup and ufw. I am positively certain that for an earlier version of docker you actually had to setup the firewall yourself. So that's just the typical documentation rot/erosion. But just like you, I also didn't have the luxury to setup this server "the proper way" and take the time to read documentation in detail. I'm using docker for years and was not aware about this ufw issue. I just kept setting up my firewalls trusting that there is no reason to check if meanwhile somebody else did the job without telling me. They usually don't tell you when they break things or don't do the job - in my experience. :-)

I'm going to write something if I find a proper place. But I guess you should be aware of this anyway if you are not already.

**Comment by polarathene** [comment]
> I hope I didn't come across as complaining about this, would have been unintentional.

No you didn't, I'm just tired :sweat_smile: 

- Clarified UFW only mentioned from community contribution and that we had a clear disclaimer pointing that out.
- Encouraged contributing better docs regarding usage with UFW, as you've detailed here.
- Explained that maintainers of DMS are limited for time, so we generally defer to community for improvements to the docs.


> But that was not meant as critique or pushing in your direction. You're doing a great job.

None taken :+1: 

I think it's great that you took the time to open an issue and provide all the detail that you have. It's just not something any maintainer is likely able to direct time towards unfortunately.

> I'm using docker for years and was not aware about this ufw issue.

I had no idea either! :open_mouth: 

This sort of thing has been a common experience since I joined as a maintainer here and get more familiar with other parts of the project or look into bug reports, it's not just the docs unfortunately :sweat_smile: 

That said, we have investigated plenty of the actual logic in the project, and addressed many of those issues. There's a tonne of niche knowledge involved though. Sometimes we'll improve on the docs if it's related to other tasks we're working on, I've done that a few times for the SSL docs.

---

> I'm going to write something if I find a proper place.

Sorry, not quite sure what you meant here. You closed the issue, are you still planning on contributing to our docs?

> But I guess you should be aware of this anyway if you are not already.

I wasn't, but thanks to your issue here I am, and at least it'll be more visible to users that search for UFW / firewall :grin:

**Comment by mutech** [comment]
> > I'm going to write something if I find a proper place.
> 
> Sorry, not quite sure what you meant here. You closed the issue, are you still planning on contributing to our docs?

Yes, but since it involves cloning/PRs etc, I'm deferring that until I found a solution for this docker/ufw thing and just offer that as a contribution instead of the docu update, I guess that adds more value. For the time being the issue is searchable if someone else stumbles over this.

> I wasn't, but thanks to your issue here I am, and at least it'll be more visible to users that search for UFW / firewall grin

Yeah, that was the idea ;-)

I opened another ticket (also closed). I am thinking about setting up the mail server differently. If I do that with docker-mailserver, I probably would need some directions. If you're interested to keep that in the scope of this project, let's have a chat. This is basically about splitting the MTA part from the  imap part, use biredctional replication & encryption offered by dovecot to have one on-premise and one online instance replicating each other and two or more mta's feeding the dovecots. As I understand it, docker-mailserver is not ideal for this use case, but I also didn't find an alternative. Would like to contribute that, just don't know how to best go about it and don't know if this is attractive enough for you guys to accept a PR.

**Comment by polarathene** [comment]
> This is basically about splitting the MTA part from the imap part, use biredctional replication & encryption offered by dovecot to have one on-premise and one online instance replicating each other and two or more mta's feeding the dovecots.
> 
> As I understand it, docker-mailserver is not ideal for this use case, but I also didn't find an alternative.

`docker-mailserver` is focused on a single image solution. I do understand the appeal of using multiple images / containers for splitting up the services though, where there is alternative dockerized mail-server projects (_Mailu, mail-in-a-box, mailcow, etc_) which I believe is better handled by those.

What we do tend to offer is an `ENABLE_<SERVICE>` ENV to internally start services like Postfix or Dovecot (_`SMTP_ONLY=1` disables Dovecot, most others use the `ENABLE_` convention_). That tends to allow you to run those with your own external service / container instead, sometimes needing to adjust config within `docker-mailserver`, which we do support.

In the past a user contributed a way to disable Dovecot so that they could use an alternative (Kopano IIRC), along with different SASL auth provider (LDAP instead of Dovecot, via SASLAuthD). Which might enable you to partially achieve your goal :+1: 

I'm not opposed to a slimmer `:core` image that is more focused on catering to the `SMTP_ONLY=1` use-case. This would be a bit of work presently as it'dl ikely be approached as an early build stage in the same Dockerfile. There is work slowly to better isolate features / services so that customizing the image like this would eventually be easier to approach.

---

> Would like to contribute that, just don't know how to best go about it and don't know if this is attractive enough for you guys to accept a PR.

I think it might be difficult to get other maintainers onboard with. Usually we advise community to add to the docs if they want to share improvements / features that aren't in heavy demand from our users.

You can open an issue proposing it and we can keep it open to see if it gets any traction from other users showing interest, otherwise it's not likely to be something we'd want the burden of maintaining (_LDAP is popular with our users, but no maintainer has experience with it that it's problematic to continue supporting_).

---

Regarding encryption, I do have Dovecot encrypted mailboxes support almost ready and hope to PR that some time this month.

There has been a user in the past that also requested support for replication I think, but it lacked interest from the community to pursue.

**Comment by mutech** [comment]
I think splitting the container in the scope of a docker-compose file should be relatively easy since at least on the file system level you can always share volumes and similarly you also have the docker networking for the rest.

But I can see that once you start modularizing the thing, docker-mailserver might very quickly loose it's core quality, the ease of setting it up.

I don't know what to think about it myself. On one hand I want to stay with this solution since it worked so well, on the other hand I find that replication model very charming and it just doesn't seem to fit well. Sure you can disable parts of mailserver, but what's the point to install all these parts with their own complexities if you then don't use them.

I looked at some of the other projects you mentioned, but none of them implements my setup out of the box and it looks like it would be just as alien in their context as in mailserver. I might be wrong, didn't invest a lot of time on this yet though. I always tend to have use cases nobody else has. That's frustrating ;-)

**Comment by polarathene** [comment]
> I would expect docker not to do it's own firewalling

I've looked into this now. When you explicitly request Docker to publish ports from a container and bind them to a host interface (default `0.0.0.0`, aka all interfaces), it will configure the firewall accordingly. This is not surprising.

Likewise with UFW (_which is a frontend to managing a firewall via `iptables` or `nftables_), it will manage rules.

Dockers docs on `iptables` mention at the bottom that there is integration with FirewallD available. This has the same issue by default, but places docker networks into a "docker" zone that can be more easily managed by FirewallD apparently.

> While this is nice, it conflicts with UFW, which does not show the firewall rules docker inserts.

The issue is that UFW is apparently oblivious to the rules Docker adds, and Dockers rules are not able to easily integrate with what UFW is doing. I am aware of Docker inserting chains `DOCKER` and `DOCKER-CHAIN`, which are applied before UFW rules, making them too late to have an affect apparently.

Try consider Docker as if you run another firewall frontend? Or think about `.bashrc` / `zshrc`, and if something you install automatically adds a snippet inside that config, it is not likely for it to understand what else is going on in the file, or if it'll be affected by a separate section in that config or break another software from functioning properly.

Part of the issue AFAIK is due to the use of `iptables`, and if Docker and UFW both used `nftables` instead of `iptables-nft` as a shim, they may have a better time co-operating? (_not an area I understand well enough unfortunately_).

> However, all exposed ports of docker-mailserver are actually allowed.

To clarify, it is published ports, not exposed ports. There is a big difference there, which is where I was originally surprised / concerned.

If you want more control (and potentially hassle if you don't know what your doing), you can tell Docker not to mess with the firewalls so much with `--iptables=false` / `{ "iptables": false }` for `/etc/docker/daemon.json`. You'll then need to manage the rules to NAT route host IP to container IPs AFAIK.

A better option is to either be explicit with `-p 127.0.0.1:25:25` / `ports: 127.0.0.1:25:25` when binding, so that the local loopback interface is used preventing public access if you don't need it. Or by changing the default bind address Docker will use when publishing ports for a container, this can be done by configuring `daemon.json` or specific networks, see the official Docker docs.

---

> if a project is based on docker and proposes to use ufw to open the firewall to support the container, then it should mention this issue.

That specific page does mention the issue now by linking here. Although the docs will get another revision in the near future.

**Comment by mutech** [comment]
Thanks for taking the time to look into this again, especially considering that the issue is closed.

I wish Docker had your work ethics, then we wouldn't have to waste our time cleaning up after these guys.

As I see it, the root of the problem is that there is no defined path for software to dynamically manage the firewall. Firewalld is a step in the right direction but both not established as default and not far reaching enough. The old open source curse, every problem has 5 solutions each of which solved two issues and creates a new problem.

Docker overreaches it's competencies however, in that it takes ownership over the firewall configuration without taking responsibility (in the sense that it doesn't do enough). UFW is popular enough on Ubuntu based systems to be seen as the canonical approach to firewall configuration. It's just not sufficient for Docker to support firewalld and silently circumvent UFW.

It's of course true that any firewall manager in the end will use nftables, but only a minority of users will actually use this mechanism directly. Another effect of Dockers approach is that software such as LXD doesn't work once Docker is fuddling around in the firewall settings, independent on whether you use UFW, while LXD uses an approach that does not negatively impact other software components. 

>  That specific page does mention the issue now by linking here. Although the docs will get another revision in the near future.

Thank you!

Big thumbs up here for your project, you're doing a fine job!
